### PR TITLE
[BPK-898] BpkInput clearable enhancements

### PR DESCRIPTION
**Pull request by shaundon**
* Switched out `clearable={true|false}` in `BpkInput` for `clearButtonMode={'never'|'while-editing'|'always'}`.
* This matches how [React Native does it](https://facebook.github.io/react-native/docs/textinput.html#clearbuttonmode).
* Defaults to `never`.
* Applied it to storybook and docs, updated tests etc.

This is something we were planning to do eventually anyway, but it's been expedited so that the date picker can be adapted to use the `always` option, as it doesn't do well with `while-editing`.

When `clearButtonMode="always"`, it'll cover up validity icons. I've cleared this with @jamesf3rguson and it's documented in the readme.

### CODE REVIEW

File: packages/bpk-component-input/package-lock.json [hunk 1]
```diff
@@ -1,17 +1,154 @@
...
+	"version": "3.2.52",
 	"lockfileVersion": 1,
+	"requires": true,
 	"dependencies": {
 		"asap": {
 			"version": "2.0.6",
 			"resolved": "https://registry.npmjs.org/asap/-/asap-2.0.6.tgz",
 			"integrity": "sha1-5QNHYR1+aQlDIIu9r+vLwvuGbUY="
 		},
+		"bpk-component-text": {
```
Comment (matthewdavidson) [inline packages/bpk-component-input/package-lock.json]: You need to revert the changes from this lockfile - this happens when you do `npm install bpk-component-text --save-dev` inside the package directory. For `bpk-*` component that we want to be symlinked by lerna, we need to manually add the entry to `package.json` and run `npm run bootstrap` - subtle difference i know, but this bit me before.

File: packages/bpk-component-input/readme.md [hunk 2]
```diff
@@ -22,30 +22,33 @@ export default () => (
     value="Edinburgh"
     onChange={() => console.log('input changed!')}
     placeholder="Country, city or airport"
+    clearButtonMode={CLEAR_BUTTON_MODES['while-editing']}
```
Comment (matthewdavidson) [inline packages/bpk-component-input/readme.md]: This is strange `CLEAR_BUTTON_MODES['while-editing']`. I know we want the keys and values matched so that flow plays nicely. But i dont think we need to use a value that has a hyphen in it? Passing strings kinda sucks right?

File: packages/bpk-mixins/src/mixins/_forms.scss [hunk 3]
```diff
@@ -132,6 +132,7 @@
   padding-right: $bpk-spacing-base;
 
   @include bpk-rtl {
+    padding-right: $bpk-input-padding-x;
```
Comment (shaundon) [inline packages/bpk-mixins/src/mixins/_forms.scss]: This is just a fix for an RTL issue I found, not strictly related to this PR.

File: packages/bpk-component-input/src/BpkInput.js [hunk 4]
```diff
@@ -78,7 +82,14 @@ const BpkInput = props => {
   }
   if (clearable) {
     classNames.push(getClassName('bpk-input--clearable'));
+    if (clearButtonMode === CLEAR_BUTTON_MODES.always) {
+      classNames.push(getClassName('bpk-input--clearable--persistent'));
+      clearButtonClassNames.push(
+        getClassName('bpk-input--clearable--persistent__clear-button'),
```
Comment (matthewdavidson) [inline packages/bpk-component-input/src/BpkInput.js]: This `classname` is non-bem. Selectors should follow:
```
.calculator {} // block 
.calculator--alternate {} // block modifier 
.calculator__definitions {} // element 
.calculator__definitions--expanded {} // element modifier
```

- `bpk-input--clearable--persistent`: This has two modifiers. I'd expect this to be a separate modifier altogether `bpk-input--persistent-clearable`
- `bpk-input--clearable--persistent__clear-button`: you cant have an element after a modifier. You need to a apply a modifier to the element in question i.e. `bpk-input__clear-button` + `bpk-input__clear-button--persistent-clearable`

File: packages/bpk-docs/src/pages/FormsPage/InputContainer.js [hunk 5]
```diff
@@ -45,8 +46,10 @@ class InputContainer extends Component {
       overrideProps = {
         value: this.state.value,
         onChange: e => this.setState({ value: e.target.value }),
-        onClear: () => this.setState({ value: '' }),
       };
+      if (FormComponent === BpkInput) {
+        overrideProps.onClear = () => this.setState({ value: '' });
+      }
```
Comment (shaundon) [inline packages/bpk-docs/src/pages/FormsPage/InputContainer.js]: Not related to this PR - we were getting an error from passing `onClear` to `BpkSelect` in the docs.

File: packages/bpk-component-input/src/BpkInput.js (hunk 4 above)
Comment (matthewdavidson) [reply packages/bpk-component-input/src/BpkInput.js]: My above comment means that you'll need to modify the logic of how you apply class names too.

File: packages/bpk-component-input/readme.md (hunk 2 above)
Comment (shaundon) [reply packages/bpk-component-input/readme.md]: Hmm yeah true. I did it to match the React Native way, but I can camelCase it instead.

File: packages/bpk-component-input/package-lock.json (hunk 1 above)
Comment (shaundon) [reply packages/bpk-component-input/package-lock.json]: Done.

File: packages/bpk-component-input/readme.md (hunk 2 above)
Comment (shaundon) [reply packages/bpk-component-input/readme.md]: Now changed!
